src/engine/QueryPlanner.cpp

```diff
@@ -2516,6 +2619,40 @@ void QueryPlanner::QueryGraph::setupGraph(
             }
           }
         }
+
+        // Add additional edges to the graph representing the connections
+        // between variables given by joins substituting cartesian product +
+        // filter.
+        for (auto& [filter, substitute] : filtersAndOptionalSubstitutes) {
+          if (!substitute.has_value()) {
+            // This filter cannot be substituted: add no edges.
+            continue;
+          }
+          std::vector<Variable> varsToBeConnected;
+          for (auto var : filter.expression_.containedVariables()) {
+            if (varToNode.contains(*var)) {
+              varsToBeConnected.push_back(*var);
+            }
+          }
+          if (varsToBeConnected.size() < 2) {
+            // There is no variables to connect, because this filter has one or
+            // zero variables.
+            continue;
+          }
+          auto first = varsToBeConnected[0];
+          for (size_t i = 1; i < varsToBeConnected.size(); i++) {
+            auto second = varsToBeConnected[i];
+            for (auto n1 : varToNode.at(first)) {
+              for (auto n2 : varToNode.at(second)) {
+                if (n1 != n2) {
+                  result[n1].insert(n2);
+                  result[n2].insert(n1);
+                }
+              }
+            }
```

Comment:
You can use `ql::views::filter` as well as `::ranges::views::cartesian::product` etc. to make this function more idiomatic.